subsys/net/lib/websocket/websocket.c

```diff
@@ -0,0 +1,855 @@
+/** @file
+ * @brief Websocket client API
+ *
+ * An API for applications to setup a websocket connections.
+ */
+
+/*
+ * Copyright (c) 2019 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(net_websocket, CONFIG_NET_WEBSOCKET_LOG_LEVEL);
+
+#include <kernel.h>
+#include <strings.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <sys/byteorder.h>
+
+#include <sys/fdtable.h>
+#include <net/net_core.h>
+#include <net/net_ip.h>
+#include <net/socket.h>
+#include <net/http_client.h>
+#include <net/websocket.h>
+
+#include <sys/byteorder.h>
+#include <base64.h>
+#include <mbedtls/sha1.h>
+
+#include "net_private.h"
+#include "sockets_internal.h"
+#include "websocket_internal.h"
+
+/* If you want to see the data that is being sent, then you can enable
+ * debugging and set the following variable to 1. It prints a lot of data
+ * so is not enabled by default.
+ */
+#define HEXDUMP_SENT_PACKETS 0
+#define HEXDUMP_RECV_PACKETS 0
+
+static struct websocket_context contexts[CONFIG_WEBSOCKET_MAX_CONTEXTS];
+
+static struct k_sem contexts_lock;
+
+extern const struct socket_op_vtable sock_fd_op_vtable;
+static const struct socket_op_vtable websocket_fd_op_vtable;
+
+static int websocket_context_ref(struct websocket_context *ctx)
+{
+	int old_rc = atomic_inc(&ctx->refcount);
+
+	return old_rc + 1;
+}
+
+static int websocket_context_unref(struct websocket_context *ctx)
+{
+	int old_rc = atomic_dec(&ctx->refcount);
+
+	if (old_rc != 1) {
+		return old_rc - 1;
+	}
+
+	return 0;
+}
+
+static inline bool websocket_context_is_used(struct websocket_context *ctx)
+{
+	NET_ASSERT(ctx);
+
+	return !!atomic_get(&ctx->refcount);
+}
+
+static struct websocket_context *websocket_get(void)
+{
+	struct websocket_context *ctx = NULL;
+	int i;
+
+	k_sem_take(&contexts_lock, K_FOREVER);
```

Comment:
Any particular reason to use semaphore here instead of mutex?

Reply:
No particular reason for that, just re-using older code that used semaphore.